src/coordination/coordinator_instance.cpp

```diff
@@ -308,6 +308,39 @@ auto CoordinatorInstance::RegisterReplicationInstance(CoordinatorClientConfig co
   return RegisterInstanceCoordinatorStatus::SUCCESS;
 }
 
+auto CoordinatorInstance::UnregisterReplicationInstance(std::string instance_name)
+    -> UnregisterInstanceCoordinatorStatus {
+  auto lock = std::lock_guard{coord_instance_lock_};
+
+  auto const name_matches = [&instance_name](ReplicationInstance const &instance) {
+    return instance.InstanceName() == instance_name;
+  };
+
+  bool is_replica{false};
+  {
+    auto inst_to_remove = std::ranges::find_if(repl_instances_, name_matches);
+    if (inst_to_remove == repl_instances_.end()) {
+      return UnregisterInstanceCoordinatorStatus::NO_INSTANCE_WITH_NAME;
+    }
+    inst_to_remove->PauseFrequentCheck();
+    is_replica = inst_to_remove->IsReplica();
+  }
+
+  if (is_replica) {
+    auto curr_main = std::ranges::find_if(repl_instances_, &ReplicationInstance::IsMain);
+    MG_ASSERT(curr_main != repl_instances_.end(), "There must be a main instance when unregistering a replica");
+    if (!curr_main->SendUnregisterReplicaRpc(instance_name)) {
+      return UnregisterInstanceCoordinatorStatus::RPC_FAILED;
+    }
+    std::erase_if(repl_instances_, name_matches);
+  } else {
+    std::erase_if(repl_instances_, name_matches);
+    TryFailover();
```

Comment:
lock is taken, I think that should be fine

Reply:
but also when we delete it, thread is destroyed. For replica callbacks I think nothing bad will happen and for main we are thinking about forbidding unregistering current main instance because if you take a look at else branch there is TryFailover call. We cannot guarantee that failover will succeed hence after unregistering we could get in a cluster state without main instance => I don't like that, unregistering instance is too rare and too simple operation to produce such result. If you want to unregister main instance, kill it and then unregister. What do you say?

Reply:
I'm thinking about the stream itself. 
We could be sending the RPC message, serializing/deserializing it, calling handlers and so on. Not sure that it safe to do. That part is not under lock.
So what happens if we destroy an instance while also communicating with it in another thread?

Reply:
I agree. We could have a stream running, and then we erase instance from our list, and since there is callback left to do, we will continue in the function finding instance and that will kill coordinator. I think that can happen, we can get on call if needed about this, it is real edge case as I see it

Reply:
All code should be written in a way that if the communication to instance fails, we can continue operating. I think it should be ok

Reply:
but good point, I will check further if this can happen